<a href="https://colab.research.google.com/github/JaimeLopezGarrido/Trabajos_Labo1/blob/main/src/ensembles/Tarea_05_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarea para el Hogar 05

Esta Tarea para el Hogar 05 se entrega el final de la cuarta clase
<br> se espera de usted que intente avanzar con los desafios propuestos y que los traiga terminados para la Clase 05 que será el miercoles 03 de septiembre

##  1. Overfitting the Public Leaderboard

Leer  https://medium.com/hmif-itb/overfitting-the-leaderboard-da25172ac62e
( 8 minutos )

## 2. Hiperparámetros del LightGBM

Los objetivos de esta tarea son:


*   Aumentar la rentabilidad de la campaña de marketing de retención proactiva de clientes.
*   Generar un mejor modelo optimizando sus hiperparámetros
*   Conceptual : investigar los mas relevantes hiperparámetros de LightGBM
*   Familiarizarse con la Bayesian Optimization, sus largos tiempos de corrida y opciones para reducirlos
*   Familiarizarse con el uso de máquinas virtuales de Google Colab
*   Ver un pipeline completo de optimización de hiperparámetros y puesta en producción

LightGBM cuenta con mas de 60 hiperparámetros, siendo posible utilizar 40 al mismo tiempo, aunque no razonable.
<br> La documentación oficial de los hiperparámetros de LightGBM es  https://lightgbm.readthedocs.io/en/latest/Parameters.html#core-parameters


Se lo alerta sobre que una Optimizacion Bayesiana lleva varias horas de corrida, y usted deberá correr VARIAS optimizaciones para descubrir cuales parámetros conviene optimizar.
<br> A pesar que la próxima clase es recien en viernes 01 de agosto, inicie la tarea con tiempo, aprenda a planificar estratégicamente sus corridas como un@ científ@  de datos.

Es necesario investigar cuales son los hiperparámetros de LightGBM que vale la pena optimizar en una Bayesian Optimization, ya que los realmente utiles son apenas un reducido subconjunto.
<br>Usted deberá investigar cuales son los hiperparámetros mas relevantes de LightGBM, su primer alternativa es preguntándole a su amigo con capacidades especiales ChatGPT o sus endogámicos familiares Claude, DeepSeek, Gemini, Grok, etc
<br> La segunda alternativa es la propia documentación de LightGBM  https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html


Adicionalmente podra buscar información como la que proveen esta diminuta muestra aleatoria de artículos ligeros:
* https://machinelearningmastery.com/light-gradient-boosted-machine-lightgbm-ensemble/
*  https://medium.com/@sarahzouinina/a-deep-dive-into-lightgbm-how-to-choose-and-tune-parameters-7c584945842e
*  https://www.kaggle.com/code/somang1418/tuning-hyperparameters-under-10-minutes-lgbm
*  https://towardsdatascience.com/beginners-guide-to-the-must-know-lightgbm-hyperparameters-a0005a812702/


<br>  La muestra anterior se brinda a modo de ejemplo, usted deberá buscar muuuuchas  fuentes adicionales de información
<br> Tenga presente que LightGBM es el estado del arte en modelado predictivo para datasets estructurado, que son el 90% del trabajo del 95% de los Data Scientists en Argentina.

El desafío de esta tarea es:
* Qué hiperparparámetros conviene optimizar?  Las recomendaciones de los artículos ligeros es siempre sensata?  Sus autores realmente hicieron experimentos o son siemplemente escritores de entretenimiento carente de base científica?
* Elegidos los hiperparámetros, cual es el  <desde, hasta> que se debe utilizar en la Bayesian Optimization ?
* Realmente vale la pena optimizar 10 o 16 hiperparámetros al mismo tiempo ?  No resulta contraproducente una búsqueda en un espacio de tal alta dimensionalidad ?



---
# Resumen Hiperparametros

LightGBM utiliza el algoritmo de crecimiento de árbol "leaf-wise" (por hoja), mientras que muchas otras herramientas populares utilizan el crecimiento "depth-wise" (por profundidad).

Parámetros Importantes para el Árbol "Leaf-wise":

- *num_leaves*  = Número máximo de hojas que cada árbol puede tener; controla complejidad.

      Este es el parámetro principal para controlar la complejidad del modelo de árbol. Una profundidad no restringida puede inducir el sobreajuste.$num\_leaves = 2^{(max\_depth)}$ para obtener el mismo número de hojas que un árbol "depth-wise".

      - Tiene que ser menor que 2^{(max\_depth)} para evitar overfitting.
      - Disminuirlo acelera el entrenamiento y reduce overfitting.

- *min_data_in_leaf* = Número mínimo de observaciones que cada hoja debe contener.
      
      Muy importante para prevenir overfitting, ya que previene hojas con muy pocos datos. Cientos o miles es suficiente para un conjunto de datos grande.
      - Si sobreajusta hay que subirlo
      - Si subajusta hay que bajarlo

- *max_depth* = Profundidad máxima de cada árbol; ≤0 significa sin límite. Disminuirlo reduce tiempo y sobreajuste.

      -1 → sin límite (LightGBM lo decide por num_leaves)

      3–8 → árboles pequeños y conservadores

      8–15 → árboles medianos

      > 15 → árboles muy complejos, riesgo de sobreajuste



---
# Arboles, menos profundos, mas rapidos, pero menos precisos.


- Disminuir *max_depth*: Reducir la profundidad máxima del árbol para limitar el crecimiento y el tiempo de entrenamiento.
- Disminuir *num_leaves*: Reducir el número máximo de nodos hoja por árbol, controlando la complejidad y disminuyendo el tiempo de entrenamiento.
- Aumentar *min_gain_to_split*: Incrementar la ganancia mínima requerida para una división, evitando divisiones de mejora muy pequeña y acortando el entrenamiento.
- Aumentar *min_data_in_leaf* y *min_sum_hessian_in_leaf*: Aumentar estos valores para prevenir la creación de nodos de árbol muy específicos (sobreajuste) y reducir el tiempo de entrenamiento.

---

# Entrenar menos árboles
*num_iterations*

- Número de árboles o rondas de boosting.

- Valores típicos: 500–2000.

- Menos iteraciones → menos tiempo de entrenamiento.

- Si se reduce num_iterations, conviene subir learning_rate (por ejemplo, de 0.03 a 0.1).

*learning_rate*

- Cuánto se ajusta el modelo en cada iteración.

- Valores típicos: 0.01–0.1.

- No afecta el tiempo de entrenamiento, pero sí la precisión.

*early_stopping_round*

- Detiene el entrenamiento si no mejora tras X iteraciones.

- Ejemplo: early_stopping_round = 50.

- Requiere usar un set de validación.

- Permite ahorrar tiempo sin perder rendimiento.

---

# Menos divisiones por nodo
*feature_pre_filter*

Si está en TRUE (por defecto), LightGBM filtra variables con poca información antes de entrenar. Mantenerlo en TRUE para acelerar.

*max_bin*

- Cantidad máxima de bins por variable continua.

- Menor valor = entrenamiento más rápido pero menos preciso.

- Valores típicos: 31–255.

- Reducirlo acorta el tiempo de entrenamiento.

*min_data_in_bin*

- Mínimo de datos por bin.

- Aumentarlo reduce el número de bins y acelera el proceso.

- Valor típico: 3–10.

*feature_fraction*

- Porcentaje de variables usadas en cada árbol.

- Ejemplo: feature_fraction = 0.5 usa el 50% de las variables al azar.

- Reducirlo mejora velocidad y ayuda a evitar sobreajuste.

- Valores típicos: 0.5–0.9.

*max_cat_threshold*

- Número máximo de combinaciones evaluadas para variables categóricas.

- Reducirlo disminuye tiempo de cómputo.

- Valores comunes: 16–64.

---


# Usar menos datos por iteración
*bagging_fraction* y *bagging_freq*

Permiten usar una muestra aleatoria del dataset en cada iteración.

Ejemplo:

- bagging_fraction = 0.75 → usa el 75% de los datos.

- bagging_freq = 5 → cambia la muestra cada 5 iteraciones.

Reducir bagging_fraction acelera el entrenamiento y puede mejorar generalización.

---

# Para una Mejor Precisión

Para maximizar la precisión del modelo (generalmente a costa de un mayor tiempo de entrenamiento):

- Usar un max_bin grande (puede ser más lento).

- Usar un learning_rate pequeño con un num_iterations grande.

- Usar un num_leaves grande (puede causar sobreajuste).

- Usar datos de entrenamiento más grandes.

- Probar el booster dart.

**Cómo Lidiar con el Sobreajuste**

Para ayudar al modelo a generalizar mejor y evitar el sobreajuste:

- Usar un max_bin pequeño.

- Usar un num_leaves pequeño.

- Usar min_data_in_leaf y min_sum_hessian_in_leaf.

- Usar bagging (agregación de bootstrap) configurando bagging_fraction y bagging_freq.

- Usar submuestreo de características (feature sub-sampling) configurando feature_fraction.

- Usar datos de entrenamiento más grandes.

- Probar lambda_l1, lambda_l2 y min_gain_to_split para regularización.

- Probar max_depth para evitar el crecimiento de árboles profundos.

- Probar el algoritmo extra_trees.

- Probar aumentar path_smooth.

#### 2.1  Seteo del ambiente en Google Colab

Esta parte se debe correr con el runtime en Python3
<br>Ir al menu, Runtime -> Change Runtime Type -> Runtime type ->  **Python 3**

Conectar la virtual machine donde esta corriendo Google Colab con el  Google Drive, para poder tener persistencia de archivos

In [1]:
# primero establecer el Runtime de Python 3
from google.colab import drive
drive.mount('/content/.drive')

Mounted at /content/.drive


Para correr la siguiente celda es fundamental en Arranque en Frio haber copiado el archivo kaggle.json al Google Drive, en la carpeta indicada en el instructivo

<br>los siguientes comando estan en shell script de Linux
*   Crear las carpetas en el Google Drive
*   "instalar" el archivo kaggle.json desde el Google Drive a la virtual machine para que pueda ser utilizado por la libreria  kaggle de Python
*   Bajar el  **dataset_pequeno**  al  Google Drive  y tambien al disco local de la virtual machine que esta corriendo Google Colab



In [2]:
%%shell

mkdir -p "/content/.drive/My Drive/labo1"
mkdir -p "/content/buckets"
ln -s "/content/.drive/My Drive/labo1" /content/buckets/b1

mkdir -p ~/.kaggle
cp /content/buckets/b1/kaggle/kaggle.json  ~/.kaggle
chmod 600 ~/.kaggle/kaggle.json


mkdir -p /content/buckets/b1/exp
mkdir -p /content/buckets/b1/datasets
mkdir -p /content/datasets



archivo_origen="https://storage.googleapis.com/open-courses/austral2025-af91/dataset_pequeno.csv"
archivo_destino="/content/datasets/dataset_pequeno.csv"
archivo_destino_bucket="/content/buckets/b1/datasets/dataset_pequeno.csv"

if ! test -f $archivo_destino_bucket; then
  wget  $archivo_origen  -O $archivo_destino_bucket
fi


if ! test -f $archivo_destino; then
  cp  $archivo_destino_bucket  $archivo_destino
fi


### 2.2 Optimizacion Hiperparámetros

Esta parte se debe correr con el runtime en lenguaje R Ir al menu, Runtime -> Change Runtime Type -> Runtime type -> R

### 2.2.1 Inicio

limpio el ambiente de R

In [1]:
format(Sys.time(), "%a %b %d %X %Y")

[1] "Mon Nov 03 02:27:41 PM 2025"

In [2]:
# limpio la memoria
rm(list=ls(all.names=TRUE)) # remove all objects
gc(full=TRUE, verbose=FALSE) # garbage collection

used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  660381 35.3 1454462    77.7 1454462  77.7
Vcells 1226627  9.4 8388608    64.0 1975128  15.1

### 2.2.2 Carga de Librerias

In [3]:
# cargo las librerias que necesito
require("data.table")
require("parallel")

if( !require("primes") ) install.packages("primes")
require("primes")

if( !require("utils") ) install.packages("utils")
require("utils")

if( !require("rlist") ) install.packages("rlist")
require("rlist")

if( !require("yaml")) install.packages("yaml")
require("yaml")

if( !require("lightgbm") ) install.packages("lightgbm")
require("lightgbm")

if( !require("DiceKriging") ) install.packages("DiceKriging")
require("DiceKriging")

if( !require("mlrMBO") ) install.packages("mlrMBO")
require("mlrMBO")

Loading required package: data.table

Loading required package: parallel

Loading required package: primes

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘primes’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Loading required package: primes

Loading required package: rlist

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘rlist’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘XML’


Loading required package: rlist

Loading required package: yaml

Loading required package: lightgbm

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘lightgbm’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Loading requ

### 2.2.3 Definicion de Parametros

aqui debe cargar SU semilla primigenia
<br>recuerde cambiar el numero de experimento en cada corrida nueva

In [4]:
PARAM <- list()

In [5]:

PARAM$experimento <- 4
PARAM$semilla_primigenia <- 100007


In [6]:
PARAM$kaggle$competencia <- "labo-i-2025-virtual-analista-sr"
PARAM$kaggle$cortes <- seq(10000, 12000, by= 500)

In [7]:
# un undersampling de 0.1  toma solo el 10% de los CONTINUA
# undersampling de 1.0  implica tomar TODOS los datos

PARAM$trainingstrategy$undersampling <- 0.5

In [8]:
# Parametros LightGBM

PARAM$hyperparametertuning$xval_folds <- 5

# parametros fijos del LightGBM que se pisaran con la parte variable de la BO
PARAM$lgbm$param_fijos <-  list(
  boosting= "gbdt", # puede ir  dart  , ni pruebe random_forest
  objective= "binary",
  metric= "auc",
  first_metric_only= FALSE,
  boost_from_average= TRUE,
  feature_pre_filter= FALSE,
  force_row_wise= TRUE, # para reducir warnings
  verbosity= -100,

  seed= PARAM$semilla_primigenia,

  max_depth= -1L, # -1 significa no limitar,  por ahora lo dejo fijo
  min_gain_to_split= 0, # min_gain_to_split >= 0
  min_sum_hessian_in_leaf= 0.001, #  min_sum_hessian_in_leaf >= 0.0
  lambda_l1= 0.0, # lambda_l1 >= 0.0
  lambda_l2= 0.0, # lambda_l2 >= 0.0
  max_bin= 31L, # lo debo dejar fijo, no participa de la BO

  bagging_fraction= 1.0, # 0.0 < bagging_fraction <= 1.0
  pos_bagging_fraction= 1.0, # 0.0 < pos_bagging_fraction <= 1.0
  neg_bagging_fraction= 1.0, # 0.0 < neg_bagging_fraction <= 1.0
  is_unbalance= FALSE, #
  scale_pos_weight= 1.0, # scale_pos_weight > 0.0

  drop_rate= 0.1, # 0.0 < neg_bagging_fraction <= 1.0
  max_drop= 50, # <=0 means no limit
  skip_drop= 0.5, # 0.0 <= skip_drop <= 1.0

  extra_trees= FALSE,

  num_iterations= 1200,
  learning_rate= 0.02,
  feature_fraction= 0.5,
  num_leaves= 750,
  min_data_in_leaf= 5000
)


Aqui se definen los hiperparámetros de LightGBM que participan de la Bayesian Optimization
<br> si es un numero entero debe ir  makeIntegerParam
<br> si es un numero real (con decimales) debe ir  makeNumericParam
<br> es muy importante leer cuales son un lower y upper  permitidos y ademas razonables

In [29]:
# Aqui se cargan los bordes de los hiperparametros de la BO
#PARAM$hypeparametertuning$hs <- makeParamSet(
#  makeNumericParam("learning_rate", lower = 0.01, upper = 0.05),
#  makeIntegerParam("num_leaves", lower = 16L, upper = 128L),
#  makeNumericParam("feature_fraction", lower = 0.5, upper = 0.9),
#  makeNumericParam("bagging_fraction", lower = 0.7, upper = 1.0),
#  makeIntegerParam("min_data_in_leaf", lower = 20L, upper = 200L),
#  makeNumericParam("lambda_l1", lower = 0, upper = 10),
#  makeNumericParam("lambda_l2", lower = 0, upper = 10),
#  makeNumericParam("min_gain_to_split", lower = 0, upper = 0.5),
#  makeIntegerParam("max_depth", lower = -1L, upper = 10L)
#)

PARAM$hypeparametertuning$hs <- makeParamSet(
  makeIntegerParam("num_leaves", lower = 16L, upper = 512L),
  makeIntegerParam("max_depth", lower = 3L, upper = 15L),
  makeNumericParam("learning_rate", lower = 1e-2, upper = 0.3),
  makeIntegerParam("n_estimators", lower = 200L, upper = 3000L),
  makeIntegerParam("min_data_in_leaf", lower = 10L, upper = 200L),
  makeNumericParam("feature_fraction", lower = 0.5, upper = 1.0),
  makeNumericParam("bagging_fraction", lower = 0.5, upper = 1.0),
  makeIntegerParam("bagging_freq", lower = 1L, upper = 10L),
  makeNumericParam("lambda_l1", lower = 1e-4, upper = 10.0),
  makeNumericParam("lambda_l2", lower = 1e-4, upper = 10.0),
  makeNumericParam("min_gain_to_split", lower = 0.0, upper = 1.0),
  makeNumericParam("scale_pos_weight", lower = 50, upper = 150)
)

A mayor cantidad de hiperparámetros, se debe aumentar las iteraciones de la Bayesian Optimization
<br> 30 es un valor muy tacaño, pero corre rápido
<br> deberia partir de 50, alcanzando los 100 si se dispone de tiempo

In [30]:
PARAM$hyperparametertuning$iteraciones <- 35 # iteraciones bayesianas

### 2.2.4  Preprocesamiento

In [31]:
# carpeta de trabajo

setwd("/content/buckets/b1/exp")
experimento_folder <- paste0("HT", PARAM$experimento)
dir.create(experimento_folder, showWarnings=FALSE)
setwd( paste0("/content/buckets/b1/exp/", experimento_folder ))

In [32]:
# lectura del dataset

dataset <- fread("/content/datasets/dataset_pequeno.csv")

In [33]:
dataset_train <- dataset[foto_mes %in% c(202107)]

In [34]:
# paso la clase a binaria que tome valores {0,1}  enteros
#  BAJA+1 y BAJA+2  son  1,   CONTINUA es 0

dataset_train[,
  clase01 := ifelse(clase_ternaria %in% c("BAJA+2","BAJA+1"), 1L, 0L)
]

In [35]:
# defino los datos que forma parte del training
# aqui se hace el undersampling de los CONTINUA
# notar que para esto utilizo la SEGUNDA semilla

set.seed(PARAM$semilla_primigenia, kind = "L'Ecuyer-CMRG")
dataset_train[, azar := runif(nrow(dataset_train))]
dataset_train[, training := 0L]

dataset_train[
  foto_mes %in% c(202107) &
    (azar <= PARAM$trainingstrategy$undersampling | clase_ternaria %in% c("BAJA+1", "BAJA+2")),
  training := 1L
]

In [36]:
# los campos que se van a utilizar

campos_buenos <- setdiff(
  colnames(dataset_train),
  c("clase_ternaria", "clase01", "azar", "training")
)

In [37]:
# dejo los datos en el formato que necesita LightGBM

dtrain <- lgb.Dataset(
  data= data.matrix(dataset_train[training == 1L, campos_buenos, with= FALSE]),
  label= dataset_train[training == 1L, clase01],
  free_raw_data= FALSE
)

nrow(dtrain)
ncol(dtrain)

[1] 83029

[1] 154

2.2.5 Configuracion Bayesian Optimization

In [38]:
# En el argumento x llegan los parmaetros de la bayesiana
#  devuelve la AUC en cross validation del modelo entrenado

EstimarGanancia_AUC_lightgbm <- function(x) {

  # x pisa (o agrega) a param_fijos
  param_completo <- modifyList(PARAM$lgbm$param_fijos, x)

  # entreno LightGBM
 modelocv <- lgb.cv(
  data= dtrain,
  nfold= PARAM$hyperparametertuning$xval_folds,
  stratified= TRUE,
  param= param_completo
)

  # obtengo la ganancia
  AUC <- modelocv$best_score

  # hago espacio en la memoria
  rm(modelocv)
  gc(full= TRUE, verbose= FALSE)

  message(format(Sys.time(), "%a %b %d %X %Y"), " AUC ", AUC)

  return(AUC)
}

In [40]:
# Aqui comienza la configuracion de la Bayesian Optimization

# en este archivo quedan la evolucion binaria de la BO
kbayesiana <- "bayesiana.RDATA"

funcion_optimizar <- EstimarGanancia_AUC_lightgbm # la funcion que voy a maximizar

configureMlr(show.learner.output= FALSE)

# configuro la busqueda bayesiana,  los hiperparametros que se van a optimizar
# por favor, no desesperarse por lo complejo

obj.fun <- makeSingleObjectiveFunction(
  fn= funcion_optimizar, # la funcion que voy a maximizar
  minimize= FALSE, # estoy Maximizando la ganancia
  noisy= TRUE,
  par.set= PARAM$hypeparametertuning$hs, # definido al comienzo del programa
  has.simple.signature= FALSE # paso los parametros en una lista
)

# cada 600 segundos guardo el resultado intermedio
ctrl <- makeMBOControl(
  save.on.disk.at.time= 600, # se graba cada 600 segundos
  save.file.path= kbayesiana
) # se graba cada 600 segundos

# indico la cantidad de iteraciones que va a tener la Bayesian Optimization
ctrl <- setMBOControlTermination(
  ctrl,
  iters= PARAM$hyperparametertuning$iteraciones
) # cantidad de iteraciones

# defino el método estandar para la creacion de los puntos iniciales,
# los "No Inteligentes"
ctrl <- setMBOControlInfill(ctrl, crit= makeMBOInfillCritEI())

# establezco la funcion que busca el maximo
surr.km <- makeLearner(
  "regr.km",
  predict.type= "se",
  covtype= "matern3_2",
  control= list(trace= TRUE)
)




2.2.6 Corrida Bayesian Optimization

In [41]:
# inicio la optimizacion bayesiana, retomando si ya existe
# es la celda mas lenta de todo el notebook

if (!file.exists(kbayesiana)) {
  bayesiana_salida <- mbo(obj.fun, learner= surr.km, control= ctrl)
} else {
  bayesiana_salida <- mboContinue(kbayesiana) # retomo en caso que ya exista
}

Computing y column(s) for design. Not provided.

Mon Nov 03 03:03:57 PM 2025 AUC 0.909532539810231

Mon Nov 03 03:07:31 PM 2025 AUC 0.921799154766002

Mon Nov 03 03:09:35 PM 2025 AUC 0.919574682168755

Mon Nov 03 03:13:50 PM 2025 AUC 0.919253964641665

Mon Nov 03 03:16:57 PM 2025 AUC 0.927627336780844

Mon Nov 03 03:19:24 PM 2025 AUC 0.90143238504341

Mon Nov 03 03:22:24 PM 2025 AUC 0.916929018814496

Mon Nov 03 03:24:39 PM 2025 AUC 0.905022142983919

Mon Nov 03 03:27:33 PM 2025 AUC 0.921449513560722

Mon Nov 03 03:29:29 PM 2025 AUC 0.90356157964467

Mon Nov 03 03:31:53 PM 2025 AUC 0.913971922950486

Mon Nov 03 03:36:13 PM 2025 AUC 0.916220851194246

Mon Nov 03 03:37:52 PM 2025 AUC 0.910608423515439

Mon Nov 03 03:39:49 PM 2025 AUC 0.911542345603073

Mon Nov 03 03:42:05 PM 2025 AUC 0.903560615389034

Mon Nov 03 03:43:57 PM 2025 AUC 0.913565715845656

Mon Nov 03 03:47:57 PM 2025 AUC 0.922743383475866

Mon Nov 03 03:51:39 PM 2025 AUC 0.915343289266232

Mon Nov 03 03:53:40 PM 2025 AUC 0.9

In [42]:

tb_bayesiana <- as.data.table(bayesiana_salida$opt.path)
colnames( tb_bayesiana)

[1] "num_leaves"        "max_depth"         "learning_rate"    
 [4] "n_estimators"      "min_data_in_leaf"  "feature_fraction" 
 [7] "bagging_fraction"  "bagging_freq"      "lambda_l1"        
[10] "lambda_l2"         "min_gain_to_split" "scale_pos_weight" 
[13] "y"                 "dob"               "eol"              
[16] "error.message"     "exec.time"         "ei"               
[19] "error.model"       "train.time"        "prop.type"        
[22] "propose.time"      "se"                "mean"

In [43]:
# almaceno los resultados de la Bayesian Optimization
# y capturo los mejores hiperparametros encontrados

tb_bayesiana <- as.data.table(bayesiana_salida$opt.path)

tb_bayesiana[, iter := .I]

# ordeno en forma descendente por AUC = y
setorder(tb_bayesiana, -y)

# grabo para eventualmente poder utilizarlos en OTRA corrida
fwrite( tb_bayesiana,
  file= "BO_log.txt",
  sep= "\t"
)

# los mejores hiperparámetros son los que quedaron en el registro 1 de la tabla
PARAM$out$lgbm$mejores_hiperparametros <- tb_bayesiana[
  1, # el primero es el de mejor AUC
  setdiff(colnames(tb_bayesiana),
    c("y","dob","eol","error.message","exec.time","ei","error.model",
      "train.time","prop.type","propose.time","se","mean","iter")),
  with= FALSE
]


PARAM$out$lgbm$y <- tb_bayesiana[1, y]


In [44]:
write_yaml( PARAM, file="PARAM.yml")

In [45]:
print(PARAM$out$lgbm$mejores_hiperparametros)
print(PARAM$out$lgbm$y)

   num_leaves max_depth learning_rate n_estimators min_data_in_leaf
        <int>     <int>         <num>        <int>            <int>
1:        179        15    0.01071297          253               96
   feature_fraction bagging_fraction bagging_freq lambda_l1 lambda_l2
              <num>            <num>        <int>     <num>     <num>
1:        0.9778362         0.879424            4  9.074415  7.236361
   min_gain_to_split scale_pos_weight
               <num>            <num>
1:          0.395588         79.35603
[1] 0.929843


# Cuarta Corrida: "GPT" Profesor

Pruebo corrida de chat con chat gpt que paso el profesor por Zulip

PARAM$hypeparametertuning$hs <- makeParamSet(
  makeIntegerParam("num_leaves", lower = 16L, upper = 512L),
  makeIntegerParam("max_depth", lower = 3L, upper = 15L),
  makeNumericParam("learning_rate", lower = 1e-2, upper = 0.3),
  makeIntegerParam("n_estimators", lower = 200L, upper = 3000L),
  makeIntegerParam("min_data_in_leaf", lower = 10L, upper = 200L),
  makeNumericParam("feature_fraction", lower = 0.5, upper = 1.0),
  makeNumericParam("bagging_fraction", lower = 0.5, upper = 1.0),
  makeIntegerParam("bagging_freq", lower = 1L, upper = 10L),
  makeNumericParam("lambda_l1", lower = 1e-4, upper = 10.0),
  makeNumericParam("lambda_l2", lower = 1e-4, upper = 10.0),
  makeNumericParam("min_gain_to_split", lower = 0.0, upper = 1.0),
  makeNumericParam("scale_pos_weight", lower = 50, upper = 150),
  makeDiscreteParam("boosting_type", values = c("gbdt", "dart"))
)

    boosting: 'gbdt'
    objective: 'binary'
    metric: 'auc'
    first_metric_only: FALSE
    boost_from_average: TRUE
    feature_pre_filter: FALSE
    force_row_wise: TRUE
    verbosity: -100
    seed: 100007
    max_depth: 15
    min_gain_to_split: 0.39558796
    min_sum_hessian_in_leaf: 0.001
    lambda_l1: 9.0744155
    lambda_l2: 7.2363611
    max_bin: 31
    bagging_fraction: 0.879424
    pos_bagging_fraction: 1
    neg_bagging_fraction: 1
    is_unbalance: FALSE
    scale_pos_weight: 79.35603169
    drop_rate: 0.1
    max_drop: 50
    skip_drop: 0.5
    extra_trees: FALSE
    num_iterations: 1200
    learning_rate: 0.01071297
    feature_fraction: 0.9778362
    num_leaves: 179
    min_data_in_leaf: 96
    n_estimators: 253
    bagging_freq: 4

    Kaggle: 57.407

## 2.3  Produccion

### Final Training
Construyo el modelo final, que es uno solo, no hace ningun tipo de particion < training, validation, testing>]

In [46]:
setwd("/content/buckets/b1/exp")
experimento <- paste0("exp", PARAM$experimento)
dir.create(experimento, showWarnings= FALSE)
setwd( paste0("/content/buckets/b1/exp/", experimento ))

#### Final Training Dataset

Aqui esta la gran decision de en qué meses hago el Final Training
<br> debo utilizar los mejores hiperparámetros que encontré en la optimización bayesiana

In [47]:
# clase01
dataset[, clase01 := ifelse(clase_ternaria %in% c("BAJA+1", "BAJA+2"), 1L, 0L)]

In [48]:
dataset_train <- dataset[foto_mes %in% c(202107)]

In [49]:
# dejo los datos en el formato que necesita LightGBM

dtrain <- lgb.Dataset(
  data= data.matrix(dataset_train[, campos_buenos, with= FALSE]),
  label= dataset_train[, clase01]
)

#### Final Training Hyperparameters

In [50]:
param_final <- modifyList(PARAM$lgbm$param_fijos,
  PARAM$out$lgbm$mejores_hiperparametros)

param_final

$boosting
[1] "gbdt"

$objective
[1] "binary"

$metric
[1] "auc"

$first_metric_only
[1] FALSE

$boost_from_average
[1] TRUE

$feature_pre_filter
[1] FALSE

$force_row_wise
[1] TRUE

$verbosity
[1] -100

$seed
[1] 100007

$max_depth
[1] 15

$min_gain_to_split
[1] 0.395588

$min_sum_hessian_in_leaf
[1] 0.001

$lambda_l1
[1] 9.074415

$lambda_l2
[1] 7.236361

$max_bin
[1] 31

$bagging_fraction
[1] 0.879424

$pos_bagging_fraction
[1] 1

$neg_bagging_fraction
[1] 1

$is_unbalance
[1] FALSE

$scale_pos_weight
[1] 79.35603

$drop_rate
[1] 0.1

$max_drop
[1] 50

$skip_drop
[1] 0.5

$extra_trees
[1] FALSE

$num_iterations
[1] 1200

$learning_rate
[1] 0.01071297

$feature_fraction
[1] 0.9778362

$num_leaves
[1] 179

$min_data_in_leaf
[1] 96

$n_estimators
[1] 253

$bagging_freq
[1] 4

#### Training
Genero el modelo final, siempre sobre TODOS los datos de  final_train, sin hacer ningun tipo de undersampling de la clase mayoritaria

In [51]:
# este punto es muy SUTIL  y será revisado en la Clase 05

param_normalizado <- copy(param_final)
param_normalizado$min_data_in_leaf <-  round(param_final$min_data_in_leaf / PARAM$trainingstrategy$undersampling)

In [52]:
  # entreno LightGBM

  modelo_final <- lgb.train(
    data= dtrain,
    param= param_normalizado
  )

In [53]:
# ahora imprimo la importancia de variables

tb_importancia <- as.data.table(lgb.importance(modelo_final))
archivo_importancia <- "impo.txt"

fwrite(tb_importancia,
  file= archivo_importancia,
  sep= "\t"
)

In [54]:
# grabo a disco el modelo en un formato para seres humanos ... ponele ...

lgb.save(modelo_final, "modelo.txt" )

### Scoring

Aplico el modelo final a los datos del futuro

In [55]:
# aplico el modelo a los datos sin clase
dfuture <- dataset[foto_mes == 202109]

# aplico el modelo a los datos nuevos
prediccion <- predict(
  modelo_final,
  data.matrix(dfuture[, campos_buenos, with= FALSE])
)

#### Tabla Prediccion

In [56]:
# tabla de prediccion

tb_prediccion <- dfuture[, list(numero_de_cliente)]
tb_prediccion[, prob := prediccion ]

# grabo las probabilidad del modelo
fwrite(tb_prediccion,
  file= "prediccion.txt",
  sep= "\t"
)

Kaggle Competition Submit

In [57]:
# genero archivos con los  "envios" mejores
# suba TODOS los archivos a Kaggle

# ordeno por probabilidad descendente
setorder(tb_prediccion, -prob)

dir.create("kaggle")

for (envios in PARAM$kaggle$cortes) {

  tb_prediccion[, Predicted := 0L] # seteo inicial a 0
  tb_prediccion[1:envios, Predicted := 1L] # marclo los primeros

  archivo_kaggle <- paste0("./kaggle/KA", PARAM$experimento, "_", envios, ".csv")

  # grabo el archivo
  fwrite(tb_prediccion[, list(numero_de_cliente, Predicted)],
    file= archivo_kaggle,
    sep= ","
  )

  # subida a Kaggle, armo la linea de comando
  comando <- "kaggle competitions submit"
  competencia <- paste("-c", PARAM$kaggle$competencia)
  arch <- paste( "-f", archivo_kaggle)

  mensaje <- paste0("-m 'envios=", envios,
  "  semilla=", PARAM$semilla_primigenia,
    "'" )

  linea <- paste( comando, competencia, arch, mensaje)

  salida <- system(linea, intern=TRUE) # el submit a Kaggle
  cat(salida, "\n")
  Sys.sleep(45)
}

Successfully submitted to LaboI 2025 virtual analista sr 
Successfully submitted to LaboI 2025 virtual analista sr 
Successfully submitted to LaboI 2025 virtual analista sr 
Successfully submitted to LaboI 2025 virtual analista sr 
Successfully submitted to LaboI 2025 virtual analista sr 


In [58]:
write_yaml( PARAM, file="PARAM.yml")

In [59]:
format(Sys.time(), "%a %b %d %X %Y")

[1] "Mon Nov 03 08:37:46 PM 2025"

Finalmente usted deberá cargar el resultado de su corrida en la Google Sheet Colaborativa,  hoja **TareaHogar-05**
<br> Siéntase libre de agregar las columnas que hagan falta a la planilla

Seguramente usted realice varias corridas de este script con distintos conjuntos de hiperparámetros, siempre cambiandole el nombre al script  y también cambiando el nombre del experimento,  deberá TODAS esas corridas en distintas lineas de la  Google Sheet Colaborativa, hoja **TareaHogar-05**

Siéntase libre de agregar columnas a la hoja **TareaHogar-05**  en caso de ser necesario.